### DISCUSSION TITLE
🆕  Prometheus Query Builder Feedback

### DISCUSSION THREAD

**Discussion by torkelo** [grafana/grafana#47693] [Observability]
In Grafana v8.5 we have added a new **Prometheus visual query builder** as a new beta feature :tada: :sparkles:  ! 

The goal of this query builder is to allow anyone to create, edit and understand PromQL queries even if you are a new Prometheus user who is not yet familiar with the syntax and mechanics of PromQL. At least that is the goal. It might take us some time to reach this goal, and your feedback could really help in shaping its future.

### What it can do

* Select metric and label filters from simple dropdowns, start with metric or start with labels as they both act as filters on each other 
* Add functions and aggregations via the operations selector 
* Range vectors are added automatically when adding a function that requires a range vector 
* Basic hint actions 
* Supports binary operations on queries 
* Parser that enabled going between the Code (text) mode and builder mode without losing changes. 
* Explain mode that breaks the query down and tries to explain what each part does. 
* A new auto legend format mode 
  * This is very basic in this first iteration. If your results only contain a single label key then the series names will contain the label values

![Screenshot from 2022-04-13 13-42-44](https://user-images.githubusercontent.com/10999/163172777-9d94f000-5195-4ca7-a192-6436737c8911.png)

### Limitations

The builder is not designed to handle all possible PromQL queries. It is designed to handle the most common forms of queries. There should be a warning when going from Code (text) mode to builder mode if the current query cannot be accurately represented in the builder mode. 

### Explain mode 

The explain mode breaks the query down into all its individual parts and functions and brings in documentation and explanations for each part. 

### Query patterns 

This feature is still in a very early state. The thinking is to allow easy access and exploration of common best practice query patterns. Query patterns are currently designed to only replace the operations (functions) of the query but leave your metric and label filters intact. If you have suggestions for this feature and suggestions for good query patterns to include let us know. 

### Common issues 

* After adding a rate function the graph/data disappear completely (no data). It could be due to too low step/scrape interval. 

### Feedback

Please let us know what you think. Your feedback could help us decide what to add or change in a future iteration. :sparkles:

**Comment by Vimao** [comment]
Just started testing this, and it looks great in terms of usability and readability improvements. 

I think what would be nice is to have the description of a specific operation visible as a popup while hovering over it, instead of needing to select it and then clicking the Information button:

![image](https://user-images.githubusercontent.com/47415312/164650578-3dde3175-7000-41d5-8cab-99cdccce3c8e.png)

What i mean is having this popup before selecting an operation, not after.

Second great thing would be an option to add custom query patterns. This functionality just begs to be used as a sort of library for the best prometheus tricks.

**Reply by torkelo** [reply]
thanks for submitting the first feedback! Yea. it's tricky with large on hover popovers, if triggered accidentally when just moving the mouse the can cover bits that you want to/planned to interact with. But if the display delay is long enough so you need to be over the info icon for some set of milliseconds (And not just accidentally passing over it). Then it could work

**Comment by metalmatze** [comment]
Oh. Just not discovered this Discussion. 
I've already opened a new issue #51628 for what I believe is a bug in the way it's handled.

**Reply by ivanahuckova** [reply]
Thank you for sharing this. The fix Pr is here: https://github.com/grafana/grafana/pull/52978

**Comment by chancez** [comment]
Tried this out today, and I could not find where I type in a raw query for quite a while. It's the _tiny_ `code` button at the top right. I didn't realize explain/code were even clickable at first. I kind of expected to be able to click/double-click in the "raw query" section where the query builder preview shows the raw query.

**Reply by i-mtz** [reply]
Yes, this is not really clear on the UI (to me at least). I also expected the "Raw query" switch to allow me to type the query as it was before. Your comment was the only hint I could find about how to go back to the original behavior.

**Comment by zackman0010** [comment]
I'm loving the query builder! Are there plans to add it to the New Alert/Recording Rule pages as well?

**Comment by luckysori** [comment]
I love the idea of query patterns, but as someone new to all this I find it hard to know what they are good for. I would welcome a short explanation on what they do, either on hover or after selecting them.

**Comment by wiardvanrij** [comment]
To be honest, I was actually a bit sceptical. Yet the experience is very smooth, it does make it easier especially for usability for a much broader audience. So kudo's for that :) 

Main feedback from me would be:
- As said before by others, the switch (builder/code) is hard to find. I would even like to suggest to place this in a much more prominent spot and always allow the option to quickly swap between builder and code. It might be useful to use the build for some parts and finish with the code part (or visa-versa). So, if the idea is to always keep this in a nice easy to use spot: 👍 
- Generally speaking, after a while, you know how to make certain queries. So the true value for me would be that the query builder allows you to make those mind boggling queries with joins, label replaces, etc. The things you (hopefully) do not use often, but when you need it, you most likely need external resources to make them. 

Call it 'sub' feedback;
- caching and/or limited the amount of metrics returned? We've got a ton of metrics and I noticed it can sometimes be a tad slow to find a metric. From a user experience point of view, it's going to be annoying to wait ~3 seconds to start a new query. It should be instant for it to be really awesome. 
- It sometimes does not allow to find any metrics when you've selected some other fields. The list then just becomes empty. I actually have not been able to reproduce it. When I can (or have more details) ill make an issue. Just figured to pass it along.

**Comment by Strunck** [comment]
The query builder is super! It makes it a lot easier. I use victoriametrics as datasource. 

A little UI inconsistency I found is the direction of the opening of the dropdown menues. Under certain circumstances they would be cut off.

To reproduce:
The browser window is resized, so that it ends around 150 pixels below the dropdown. Not all menue items are showing. The item with the name "base" is cut off. 
![CutOff](https://github.com/grafana/grafana/assets/64785859/91134c75-52ed-4226-b4dd-60c360443df7)

Then I make the browser window even smaller, so that the window ends around 40 pixel below. Now the behavior is as expected and the dropdown menu opens upwards and shows all its items.
![Upwards](https://github.com/grafana/grafana/assets/64785859/5c5a7ba9-365c-429c-8eb1-0f0fa9df871a)

**Comment by natellium** [comment]
👋  I'll be closing this Discussion. If you have any feedback you can create a [GH issue](https://github.com/grafana/grafana/issues/new/choose)
